### PR TITLE
del InvoiceUser entry when unable to locate the invoice

### PR DESCRIPTION
**Pull request by nicolasburtey**
It seems there are different states of deletion for an invoice:
- first state is the invoice is written as "cancelled" but the reference is still present in the bbolt database.
- second state is that the invoice is being pruned and the reference is no longer present in the bbolt database.

we were currently not managing the second case correctly. this PR should fix it. I'm not sure how we can test this programmatically.

I'm not sure what make the invoice transition from 1 to 2. maybe a lnd restart with db compaction?
(is that correct @alexbosworth?)

### CODE REVIEW

File: src/Lightning.ts [hunk 1]
```diff
@@ -807,8 +807,24 @@ export const LightningMixin = (superclass) =>
         // TODO: we should not log/keep secret in the logs
         this.logger.debug({ invoice, user: this.user }, "got invoice status")
       } catch (err) {
-        const error = `issue fetching invoice`
-        this.logger.warn({ err, invoice }, error)
+        const invoiceNotFound = "unable to locate invoice"
+        try {
```
Comment (bodymindarts) [inline src/Lightning.ts:811]: Why is this a `try/catch` instead of an `if/else`?

File: src/Lightning.ts [hunk 2]
```diff
@@ -807,8 +807,24 @@ export const LightningMixin = (superclass) =>
...
+          this.logger.error(
+            { err, err2, invoice },
+            "issue fetching invoice. unknown error",
+          )
+          return false
+        }
+
+        this.logger.info({ err, invoice }, invoiceNotFound)
+        try {
+          await InvoiceUser.deleteOne({ _id: hash, uid: this.user._id })
```
Comment (bodymindarts) [inline src/Lightning.ts:823]: Should this always be tried here?
I think this code block 811 - 827 could be structured clearer via explicit if/else control flow.

**Review by krtk6160 — APPROVED** [review]

**Review by bodymindarts — COMMENTED** [review]
I find it hard to follow the logic of the `try/catch` stuff here. I would advocate to re-write it with regular `if/else` control flow and only wrapping the external api calls.

That being said I understand this is some kind of emergency so perhaps we should merge and clean up later.

File: src/Lightning.ts (hunk 1 above)
Comment (nicolasburtey) [reply src/Lightning.ts:811]: I agree this should be a if/else but the reason I did a try/catch instead if because the error being returned from the lightning library I believe is loosely defined (and currently not typed). so I'm not sure there is always an `err` property at `err[2]` and and that this `err` contains a `details` property. would need to parse the alexbosworth/lightning library to make this is the case. I have had previous issue with it in the past. (it's also an pseudo emergency PR)
